main/src/ca/uwaterloo/flix/language/phase/constraintgeneration/TypingContext.scala

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2024 Matthew Lutze
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase.constraintgeneration
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.{Ast, Kind, Level, RigidityEnv, SourceLocation, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.language.phase.constraintgeneration.TypingConstraint.Provenance
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+/**
+  * The typing context is a mutable environment that tracks information during type constraint generation.
+  * It maintains a stack of information: Whenever inference enters a region, the current information is pushed
+  * onto the stack, and the information is reset. On exiting a region, a special constraint is added which incorporates
+  * all the constraints.
+  */
+class TypingContext {
+
+  /**
+    * Stores typing information relating to a particular scope.
+    *
+    * @param region  the region symbol associated with the scope (None if not in a region).
+    * @param constrs the constraints inferred for the scope.
+    */
+  private case class ScopeInfo(region: Option[Symbol.KindedTypeVarSym], constrs: ListBuffer[TypingConstraint])
```

Comment:
ScopeConstraints?

Reply:
In several places we access constr directly. I think we should at least have a addConstraint method.